MaxText/layers/linears.py

```diff
@@ -315,6 +326,69 @@ def generate_kernels(self, num_experts, base_emb_dim, mlp_dim):
     wo_kernel = jnp.asarray(wo_kernel, self.dtype)
     return w0_kernel, w1_kernel, wo_kernel
 
+  def permute(self, inputs, gate_logits, emb_dim):
+    inputs_2d = jnp.reshape(inputs, (-1, emb_dim))
+    weights, selected_experts = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    weights = jax.nn.softmax(weights.astype(self.weight_dtype), axis=-1).astype(self.dtype)
+    flatten_selected_experts = jnp.ravel(selected_experts)
+    sorted_selected_experts = jnp.argsort(flatten_selected_experts)
+    repeat_hidden_states = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
+    sorted_output = jnp.take(repeat_hidden_states, indices=sorted_selected_experts, axis=0).astype(self.dtype)
+    expert_order, expert_size = jnp.unique(flatten_selected_experts, return_counts=True, size=self.num_experts)
+    group_size = jnp.zeros(self.num_experts, dtype=jnp.int32)
+
+    expert_is_unused = expert_size == 0
+    indices = jnp.where(expert_is_unused, len(group_size), expert_order)
+    group_size = group_size.at[indices].set(jnp.where(expert_is_unused, group_size[indices], expert_size))
+
+    return sorted_output, sorted_selected_experts, weights, group_size
+
+  def unpermute(self, intermediate, inputs, sorted_selected_experts, weights):
+    unsort_output = jnp.take(intermediate, indices=jnp.argsort(sorted_selected_experts), axis=0)
+    flatten_weights = jnp.ravel(weights)
+    combined_output = jnp.multiply(unsort_output, flatten_weights[:, None])
+    groups = jnp.reshape(combined_output, (-1, self.num_experts_per_tok, combined_output.shape[1]))
+    return jnp.sum(groups, axis=1).reshape(inputs.shape).astype(self.dtype)
+  
+  def call_gmm(self, inputs, group_sizes, mlp_activation, w0_kernel, w1_kernel, wo_kernel):
+    @functools.partial(
+        shard_map.shard_map,
+        mesh=self.mesh,
+        in_specs=(
```

Comment:
Why are we choosing None for the sharding here? Can we add a comment?

Reply:
Even for single host we have generally have multiple devices (e.g. v4-8 is single host but has 4 devices that by default use FSDP in maxtext)